src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -643,55 +625,161 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+
+                // T[]? array = PrivateImplementationDetails.cachingField;
+                // if (array is not null) goto arrayNotNull;
+                _builder.EmitOpCode(ILOpCode.Ldsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitBranch(ILOpCode.Brtrue, arrayNotNullLabel);
+
+                // array = new T[elementCount];
+                // RuntimeHelpers.InitializeArray(token, array);
+                // PrivateImplementationDetails.cachingField = array;
+                _builder.EmitOpCode(ILOpCode.Pop);
+                _builder.EmitIntConstant(elementCount);
+                _builder.EmitOpCode(ILOpCode.Newarr);
+                EmitSymbolToken(arrayType.ElementType, wrappedExpression.Syntax);
+                _builder.EmitArrayBlockInitializer(data, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitOpCode(ILOpCode.Stsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+
+                // arrayNotNullLabel:
+                // new ReadOnlySpan<T>(array)
+                _builder.MarkLabel(arrayNotNullLabel);
+                _builder.EmitOpCode(ILOpCode.Newobj, 0);
+                EmitSymbolToken(rosArrayCtor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
+                return true;
+            }
+
+            // Emit: new ReadOnlySpan<ElementType>(PrivateImplementationDetails.cachingField ??= new ElementType[] { ... constants ... })
+            bool tryEmitAsCachedArrayOfConstants(BoundArrayCreation arrayCreation, ArrayTypeSymbol arrayType, TypeSymbol elementType, NamedTypeSymbol spanType, bool used, BoundExpression? inPlaceTarget, out bool avoidInPlace)
+            {
+                avoidInPlace = false;
+                var initializer = arrayCreation.InitializerOpt;
+                if (initializer == null)
+                {
+                    return false;
+                }
+
+                var initializers = initializer.Initializers;
+                if (initializers.Any(static init => init.ConstantValueOpt == null))
+                {
+                    return false;
+                }
+
+                if (!tryGetReadOnlySpanArrayCtor(arrayCreation.Syntax, out var rosArrayCtor))
+                {
+                    return false;
+                }
+
+                Debug.Assert(!elementType.IsEnumType());
+
+                ImmutableArray<ConstantValue> constants = initializers.Select(static init => init.ConstantValueOpt!).ToImmutableArray();
+
+                if (constants.IsEmpty)
+                {
+                    emitEmptyReadonlySpan(spanType, arrayCreation, used, inPlaceTarget);
+                    return true;
+                }
+
+                if (inPlaceTarget is not null)
```

Comment:
:memo: I'm not 100% sure about this part. Without it, we leave something on the stack which results in an assertion when the subsequent statement gets emitted.

Reply:
> I'm not 100% sure about this part.

When we are relying on a constructor, I think we can handle `inPlaceTarget` case the same way it is handled in ```if (specialElementType.SizeInBytes() == 1)``` above. We also should be able to handle it in ```emitAsCachedArrayFromBlob```, but we never reach it due to a similar early check. Might be worth adjusting that as well.